### PR TITLE
Add support for recursive loops

### LINKED ISSUE #3416: Feature idea: Recursive loop
# Recursive loops
## Why
While it's currently possible to loop through recursive tree-like structures in Twig, it requires creating a macro, which is slightly inconvenient and hard to get right without copy-pasting. This feature would add loop recursion as a built-in feature of Twig.

## Syntax
There are several ways the syntax could look, each with pros and cons. Here are two of them, I'm sure there are more:

### recurse keyword
```twig
<ul>
  {% for item in tree %}
    <li>
      {{ item.title }}
      {% if item.children is not empty %}
        <ul>
          {% recurse item.children }}
        </ul>
      {% endif %}
    </li>
  {% endfor %]
</ul>
```
Pros: Intuitive
Cons: Requires new keyword

### recurse function in the loop variable
```twig
<ul>
  {% for item in tree %}
    <li>
      {{ item.title }}
      {% if item.children is not empty %}
        <ul>
          {% loop.recurse(item.children) }}
        </ul>
      {% endif %}
    </li>
  {% endfor %]
</ul>
```
Pros: No new keywords
Cons: Seems like a mis-use of the loop variable; might make loops that are using the loop variable slightly less efficient unless the recursion can be optimized away if it's not used

## Implementation
An immediately invoked anonymous function that calls itself recursively, probably as an entry in the context array
Both of the previous examples would render as something along these lines:
```php
$context['loop_function'] = function($items) use (&$context) {
    foreach ($items as $item) {
        echo '<li>';
        echo $item['title'];
        if (array_key_exists('children', $item) && !empty($item['children'])) {
            echo '<ul>';
            $context['loop_function']($item['children']);
            echo '</ul>';
        }
        echo '</li>';
    }
};
echo '<ul>';
$context['loop_function']($context['tree']);
echo '</ul>';
```

## Who
I'll be happy to work on a PR if you think this feature makes sense.

### LINKED ISSUE #4144: Add tree tag to Twig 4.0
I think the tree tag for recursing data structures where the children have the same structure as the parent nodes is pretty useful.

```twig
{% tree item in menu %}
  {% if treeloop.first %}<ul>{% endif %}
    <li>
        <a href="{{ item.url }}">{{ item.name }}</a>
        {% subtree item.children %}
    </li>
  {% if treeloop.last %}</ul>{% endif %}
{% endtree %}
```

Just like a `{% for %}` loop, you can access the key of each list item:
```twig
{% tree key, item in menu %}
  <li>
    <b>Item {{ key }}</b>: {{ item.name }}
    {% subtree item.children %}
  </li>
{% endtree %}
```

Since Twig 4 is under development, would you consider adding this extension to the Twig core?

https://github.com/tacman/twig-tree-tag

I can clean up the documentation and such a bit if that helps.  I didn't write the original or even the first set of tweaks, but I did get it working with Twig 3 and the dev version of Twig 4.

### PR DESCRIPTION
**Pull request by fabpot**
Fixes #3416
Fixes #4144

### CODE REVIEW

**Comment by smnandre** [comment]
Should loop(loop.parent) be forbidden / prevented ?

(or any infinite récursion if this syntax has not the effect i anticipate)

**Comment by fabpot** [comment]
> Should loop(loop.parent) be forbidden / prevented ?
> 
> (or any infinite récursion if this syntax has not the effect i anticipate)

I don't think we need to do anything here. Sure, you can create an infinite loop, but there will always be ways to mess up.

**Comment by fabpot** [comment]
> Should loop(loop.parent) be forbidden / prevented ?
> 
> (or any infinite récursion if this syntax has not the effect i anticipate)

My first reaction was that we should not do anything (PHP does nothing to prevent such infinite loops for instance). But as this is a templating engine that can be used by end users that you don't control, I've limited the depth to 50 (50 is kind of arbitrary, but maybe a good balance between valid use cases and malicious usage of a recursive loop).

File: doc/tags/for.rst [hunk 1]
```diff
@@ -58,15 +58,17 @@ Inside of a ``for`` loop block you can access some special variables:
...
-``loop.index0``       The current iteration of the loop. (0 indexed)
+``loop.index``        The current iteration of the loop (1 indexed)
+``loop.index0``       The current iteration of the loop (0 indexed)
 ``loop.revindex``     The number of iterations from the end of the loop (1 indexed)
 ``loop.revindex0``    The number of iterations from the end of the loop (0 indexed)
 ``loop.first``        True if first iteration
 ``loop.last``         True if last iteration
 ``loop.length``       The number of items in the sequence
 ``loop.parent``       The parent context
 ``loop.cycle``        Cycle over a sequence of values
```
Comment (smnandre) [inline doc/tags/for.rst]: Should this one be removed this list, as it's not a variable but a function ? (#nitpicking)

(the only one from this list i believe)

Comment (smnandre) [reply doc/tags/for.rst]: (i tried to point a the `loop.cycle` line ... and failed miserably 😅 )

Comment (fabpot) [reply doc/tags/for.rst]: See #4159

File: src/Runtime/LoopContext.php [hunk 2]
```diff
@@ -25,6 +25,9 @@ final class LoopContext
     public function __construct(
         private LoopIterator $loop,
         private $parent,
+        private $blocks,
+        private $recurseFunc,
+        private $depth,
```
Comment (stof) [inline src/Runtime/LoopContext.php:30]: shouldn't those be typed ?

Comment (xabbuh) [reply src/Runtime/LoopContext.php:30]: see #4168

**Pull request merged**